# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: media/webrtc/signaling/src/media-conduit/AudioConduit.h
# Commit: b94935661622
# Full Hash: b949356616223660996acf39e1e353e7bbd2a60d
# Author: Dan Minor <dminor@mozilla.com>
# Date: 2018-11-03 09:44:53
# Regressor Bug: 1376873
# File Overlap Count: 4
# Description:
#   Bug 1376873 - Use Call interface in AudioConduit; r=padenot
#   
#   Differential Revision: https://phabricator.services.mozilla.com/D7441
# ==============================================================================

diff -r 6eca45c83131 -r b94935661622 media/webrtc/signaling/src/media-conduit/AudioConduit.h
--- a/media/webrtc/signaling/src/media-conduit/AudioConduit.h	Mon Mar 26 16:19:23 2018 -0400
+++ b/media/webrtc/signaling/src/media-conduit/AudioConduit.h	Tue Aug 21 13:39:53 2018 -0400
@@ -14,29 +14,14 @@
 #include "MediaConduitInterface.h"
 #include "MediaEngineWrapper.h"
 #include "RtpSourceObserver.h"
+#include "RtpPacketQueue.h"
 
 // Audio Engine Includes
 #include "webrtc/common_types.h"
 #include "webrtc/modules/audio_device/include/fake_audio_device.h"
 #include "webrtc/voice_engine/include/voe_base.h"
-#include "webrtc/voice_engine/include/voe_volume_control.h"
-#include "webrtc/voice_engine/include/voe_codec.h"
-#include "webrtc/voice_engine/include/voe_file.h"
-#include "webrtc/voice_engine/include/voe_network.h"
-#include "webrtc/voice_engine/include/voe_external_media.h"
-#include "webrtc/voice_engine/include/voe_audio_processing.h"
-#include "webrtc/voice_engine/include/voe_video_sync.h"
-#include "webrtc/voice_engine/include/voe_rtp_rtcp.h"
 #include "webrtc/voice_engine/channel_proxy.h"
 
-//Some WebRTC types for short notations
- using webrtc::VoEBase;
- using webrtc::VoENetwork;
- using webrtc::VoECodec;
- using webrtc::VoEExternalMedia;
- using webrtc::VoEAudioProcessing;
- using webrtc::VoEVideoSync;
- using webrtc::VoERTP_RTCP;
 /** This file hosts several structures identifying different aspects
  * of a RTP Session.
  */
@@ -62,18 +47,23 @@
    * APIs used by the registered external transport to this Conduit to
    * feed in received RTP Frames to the VoiceEngine for decoding
    */
-  virtual MediaConduitErrorCode ReceivedRTPPacket(const void *data, int len, uint32_t ssrc) override;
+  MediaConduitErrorCode ReceivedRTPPacket(const void *data, int len, uint32_t ssrc) override;
 
   /**
    * APIs used by the registered external transport to this Conduit to
    * feed in received RTCP Frames to the VoiceEngine for decoding
    */
-  virtual MediaConduitErrorCode ReceivedRTCPPacket(const void *data, int len) override;
+  MediaConduitErrorCode ReceivedRTCPPacket(const void *data, int len) override;
 
-  virtual MediaConduitErrorCode StopTransmitting() override;
-  virtual MediaConduitErrorCode StartTransmitting() override;
-  virtual MediaConduitErrorCode StopReceiving() override;
-  virtual MediaConduitErrorCode StartReceiving() override;
+  MediaConduitErrorCode StopTransmitting() override;
+  MediaConduitErrorCode StartTransmitting() override;
+  MediaConduitErrorCode StopReceiving() override;
+  MediaConduitErrorCode StartReceiving() override;
+
+  MediaConduitErrorCode StopTransmittingLocked();
+  MediaConduitErrorCode StartTransmittingLocked();
+  MediaConduitErrorCode StopReceivingLocked();
+  MediaConduitErrorCode StartReceivingLocked();
 
   /**
    * Function to configure send codec for the audio session
@@ -83,7 +73,7 @@
    * NOTE: This API can be invoked multiple time. Invoking this API may involve restarting
    *        transmission sub-system on the engine.
    */
-  virtual MediaConduitErrorCode ConfigureSendMediaCodec(const AudioCodecConfig* codecConfig) override;
+  MediaConduitErrorCode ConfigureSendMediaCodec(const AudioCodecConfig* codecConfig) override;
   /**
    * Function to configure list of receive codecs for the audio session
    * @param sendSessionConfig: CodecConfiguration
@@ -93,7 +83,7 @@
    * NOTE: This API can be invoked multiple time. Invoking this API may involve restarting
    *        transmission sub-system on the engine.
    */
-  virtual MediaConduitErrorCode ConfigureRecvMediaCodecs(
+  MediaConduitErrorCode ConfigureRecvMediaCodecs(
     const std::vector<AudioCodecConfig* >& codecConfigList) override;
 
   MediaConduitErrorCode
@@ -104,9 +94,9 @@
    * Register External Transport to this Conduit. RTP and RTCP frames from the VoiceEngine
    * shall be passed to the registered transport for transporting externally.
    */
-  virtual MediaConduitErrorCode SetTransmitterTransport(RefPtr<TransportInterface> aTransport) override;
+  MediaConduitErrorCode SetTransmitterTransport(RefPtr<TransportInterface> aTransport) override;
 
-  virtual MediaConduitErrorCode SetReceiverTransport(RefPtr<TransportInterface> aTransport) override;
+  MediaConduitErrorCode SetReceiverTransport(RefPtr<TransportInterface> aTransport) override;
 
   /**
    * Function to deliver externally captured audio sample for encoding and transport
@@ -124,11 +114,11 @@
    *       This ensures the inserted audio-samples can be transmitted by the conduit
    *
    */
-  virtual MediaConduitErrorCode SendAudioFrame(const int16_t speechData[],
-                                               int32_t lengthSamples,
-                                               int32_t samplingFreqHz,
-                                               uint32_t channels,
-                                               int32_t capture_time) override;
+  MediaConduitErrorCode SendAudioFrame(const int16_t speechData[],
+                                       int32_t lengthSamples,
+                                       int32_t samplingFreqHz,
+                                       uint32_t channels,
+                                       int32_t capture_time) override;
 
   /**
    * Function to grab a decoded audio-sample from the media engine for rendering
@@ -147,50 +137,57 @@
    *       This ensures the decoded samples are ready for reading and playout is enabled.
    *
    */
-   virtual MediaConduitErrorCode GetAudioFrame(int16_t speechData[],
-                                              int32_t samplingFreqHz,
-                                              int32_t capture_delay,
-                                              int& lengthSamples) override;
+   MediaConduitErrorCode GetAudioFrame(int16_t speechData[],
+                                       int32_t samplingFreqHz,
+                                       int32_t capture_delay,
+                                       int& lengthSamples) override;
 
 
   /**
    * Webrtc transport implementation to send and receive RTP packet.
    * AudioConduit registers itself as ExternalTransport to the VoiceEngine
    */
-   virtual bool SendRtp(const uint8_t* data,
-                        size_t len,
-                        const webrtc::PacketOptions& options) override;
+   bool SendRtp(const uint8_t* data,
+                size_t len,
+                const webrtc::PacketOptions& options) override;
 
   /**
    * Webrtc transport implementation to send and receive RTCP packet.
    * AudioConduit registers itself as ExternalTransport to the VoiceEngine
    */
-  virtual bool SendRtcp(const uint8_t *data,
-                        size_t len) override;
+  bool SendRtcp(const uint8_t *data,
+                size_t len) override;
 
-  virtual uint64_t CodecPluginID() override { return 0; }
-  virtual void SetPCHandle(const std::string& aPCHandle) override {}
-  MediaConduitErrorCode DeliverPacket(const void *data, int len) override
-  {
-    return kMediaConduitNoError;
-  }
+  uint64_t CodecPluginID() override { return 0; }
+  void SetPCHandle(const std::string& aPCHandle) override {}
+  MediaConduitErrorCode DeliverPacket(const void *data, int len) override;
 
-  virtual void DeleteStreams() override {}
+  void DeleteStreams() override {}
 
-  explicit WebrtcAudioConduit():
-                      mVoiceEngine(nullptr),
-                      mFakeAudioDevice(new webrtc::FakeAudioDeviceModule()),
-                      mTransportMonitor("WebrtcAudioConduit"),
-                      mTransmitterTransport(nullptr),
-                      mReceiverTransport(nullptr),
-                      mEngineTransmitting(false),
-                      mEngineReceiving(false),
-                      mChannel(-1),
-                      mDtmfEnabled(false),
-                      mCodecMutex("AudioConduit codec db"),
-                      mCaptureDelay(150),
-                      mSamples(0),
-                      mLastSyncLog(0)
+  explicit WebrtcAudioConduit(RefPtr<WebRtcCallWrapper> aCall,
+                              nsCOMPtr<nsIEventTarget> aStsThread)
+    : mFakeAudioDevice(new webrtc::FakeAudioDeviceModule())
+    , mTransportMonitor("WebrtcAudioConduit")
+    , mTransmitterTransport(nullptr)
+    , mReceiverTransport(nullptr)
+    , mCall(aCall)
+    , mRecvStreamConfig()
+    , mRecvStream(nullptr)
+    , mSendStreamConfig(this) // 'this' is stored but not  dereferenced in the constructor.
+    , mSendStream(nullptr)
+    , mRecvSSRC(0)
+    , mEngineTransmitting(false)
+    , mEngineReceiving(false)
+    , mRecvChannel(-1)
+    , mRecvChannelProxy(nullptr)
+    , mSendChannel(-1)
+    , mSendChannelProxy(nullptr)
+    , mDtmfEnabled(false)
+    , mMutex("WebrtcAudioConduit::mMutex")
+    , mCaptureDelay(150)
+    , mSamples(0)
+    , mLastSyncLog(0)
+    , mStsThread(aStsThread)
   {
   }
 
@@ -198,8 +195,10 @@
 
   MediaConduitErrorCode Init();
 
-  int GetChannel() { return mChannel; }
-  webrtc::VoiceEngine* GetVoiceEngine() { return mVoiceEngine; }
+  int GetRecvChannel() { return mRecvChannel; }
+  webrtc::VoiceEngine* GetVoiceEngine() {
+    return mCall->Call()->voice_engine();
+  }
 
   /* Set Local SSRC list.
    * Note: Until the refactor of the VoE into the call API is complete
@@ -207,10 +206,7 @@
    */
   bool SetLocalSSRCs(const std::vector<unsigned int>& aSSRCs) override;
   std::vector<unsigned int> GetLocalSSRCs() override;
-  bool SetRemoteSSRC(unsigned int ssrc) override
-  {
-    return false;
-  }
+  bool SetRemoteSSRC(unsigned int ssrc) override;
   bool UnsetRemoteSSRC(uint32_t ssrc) override
   {
     return true;
@@ -281,43 +277,40 @@
   WebrtcAudioConduit(const WebrtcAudioConduit& other) = delete;
   void operator=(const WebrtcAudioConduit& other) = delete;
 
-  //Local database of currently applied receive codecs
-  typedef std::vector<AudioCodecConfig* > RecvCodecList;
-
   //Function to convert between WebRTC and Conduit codec structures
   bool CodecConfigToWebRTCCodec(const AudioCodecConfig* codecInfo,
-                                webrtc::CodecInst& cinst);
+                                webrtc::AudioSendStream::Config& config);
 
   //Generate block size in sample lenght for a given sampling frequency
   unsigned int GetNum10msSamplesForFrequency(int samplingFreqHz) const;
 
-  // Function to copy a codec structure to Conduit's database
-  bool CopyCodecToDB(const AudioCodecConfig* codecInfo);
-
-  // Functions to verify if the codec passed is already in
-  // conduits database
-  bool CheckCodecForMatch(const AudioCodecConfig* codecInfo) const;
-  bool CheckCodecsForMatch(const AudioCodecConfig* curCodecConfig,
-                           const AudioCodecConfig* codecInfo) const;
   //Checks the codec to be applied
   MediaConduitErrorCode ValidateCodecConfig(const AudioCodecConfig* codecInfo, bool send);
 
-  //Utility function to dump recv codec database
-  void DumpCodecDB() const;
+  MediaConduitErrorCode CreateSendStream();
+  void DeleteSendStream();
+  MediaConduitErrorCode CreateRecvStream();
+  void DeleteRecvStream();
 
-  webrtc::VoiceEngine* mVoiceEngine;
+  MediaConduitErrorCode CreateChannels();
+  void DeleteChannels();
+
   UniquePtr<webrtc::FakeAudioDeviceModule> mFakeAudioDevice;
   mozilla::ReentrantMonitor mTransportMonitor;
   RefPtr<TransportInterface> mTransmitterTransport;
   RefPtr<TransportInterface> mReceiverTransport;
-  ScopedCustomReleasePtr<webrtc::VoENetwork>   mPtrVoENetwork;
-  ScopedCustomReleasePtr<webrtc::VoEBase>      mPtrVoEBase;
-  ScopedCustomReleasePtr<webrtc::VoECodec>     mPtrVoECodec;
-  ScopedCustomReleasePtr<webrtc::VoEExternalMedia> mPtrVoEXmedia;
-  ScopedCustomReleasePtr<webrtc::VoEAudioProcessing> mPtrVoEProcessing;
-  ScopedCustomReleasePtr<webrtc::VoEVideoSync> mPtrVoEVideoSync;
-  ScopedCustomReleasePtr<webrtc::VoERTP_RTCP>  mPtrVoERTP_RTCP;
-  ScopedCustomReleasePtr<webrtc::VoERTP_RTCP>  mPtrRTP;
+  ScopedCustomReleasePtr<webrtc::VoEBase> mPtrVoEBase;
+
+  const RefPtr<WebRtcCallWrapper> mCall;
+  webrtc::AudioReceiveStream::Config mRecvStreamConfig;
+  webrtc::AudioReceiveStream* mRecvStream;
+  webrtc::AudioSendStream::Config mSendStreamConfig;
+  webrtc::AudioSendStream* mSendStream;
+
+  // accessed on creation, and when receiving packets
+  Atomic<uint32_t> mRecvSSRC; // this can change during a stream!
+  RtpPacketQueue mRtpPacketQueue;
+
   //engine states of our interets
   mozilla::Atomic<bool> mEngineTransmitting; // If true => VoiceEngine Send-subsystem is up
   mozilla::Atomic<bool> mEngineReceiving;    // If true => VoiceEngine Receive-subsystem is up
@@ -331,12 +324,13 @@
   };
   AutoTArray<Processing,8> mProcessing;
 
-  int mChannel;
-  std::unique_ptr<webrtc::voe::ChannelProxy> mChannelProxy;
+  int mRecvChannel;
+  std::unique_ptr<webrtc::voe::ChannelProxy> mRecvChannelProxy;
+  int mSendChannel;
+  std::unique_ptr<webrtc::voe::ChannelProxy> mSendChannelProxy;
   bool mDtmfEnabled;
-  RecvCodecList    mRecvCodecList;
 
-  Mutex mCodecMutex; // protects mCurSendCodecConfig
+  Mutex mMutex;
   nsAutoPtr<AudioCodecConfig> mCurSendCodecConfig;
 
   // Current "capture" delay (really output plus input delay)
@@ -348,6 +342,9 @@
   uint32_t mLastSyncLog;
 
   RtpSourceObserver mRtpSourceObserver;
+
+  // Socket transport service thread. Any thread.
+  const nsCOMPtr<nsIEventTarget> mStsThread;
 };
 
 } // end namespace